File: src/CountAndSay.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class CountAndSay {
    private static String countAndSay(int number) {
        if (number == 1) {
            return "1";
        }

        String previous = countAndSay(number - 1);
        List<Pair> pairs = getFrequencies(previous);
        StringBuilder result = new StringBuilder();
        for (Pair pair : pairs) {
            result.append(pair.frequency).append(pair.character);
        }
        return result.toString();
    }

    private static List<Pair> getFrequencies(String string) {
        char current = string.charAt(0);
        int frequency = 0;
        List<Pair> result = new ArrayList<>();

        for (int index = 0 ; index < string.length() ; index++) {
            if (string.charAt(index) != current) {
                result.add(new Pair(current, frequency));
                frequency = 1;
                current = string.charAt(index);
            } else {
                frequency++;
            }
        }

        result.add(new Pair(current, frequency));
        return result;
    }

    private static class Pair {
        final char character;
        final int frequency;

        Pair(char character, int frequency) {
            this.character = character;
            this.frequency = frequency;
        }
    }
}

File: python/count_and_say.py
class Solution:
    def countAndSay(self, n: int) -> str:
        if n == 1:
            return '1'

        previous = self.countAndSay(n - 1)
        # print('previous', n - 1, previous)
        current = previous[0]
        frequency = 0
        result = ''
        for number in previous:
            if number == current:
                frequency += 1
            else:
                result += f'{frequency}{current}'
                frequency = 1
                current = number
        result += f'{frequency}{current}' if frequency > 0 else ''
        return result
